File: cellpaint_database/database_loader.R
library(DBI)
library(RSQLite)
library(jpeg)
library(png)
library(tiff)
library(magrittr)

DatabaseLoader <- setRefClass(
  "DatabaseLoader",
  fields = list(db_file = "character", data_root = "character", conn = "ANY", table_name = "character"),
  methods = list(
    initialize = function(db_file, data_root) {
      db_file <<- db_file
      data_root <<- data_root
      conn <<- dbConnect(SQLite(), db_file)
      
      tables <- dbGetQuery(conn, "SELECT name FROM sqlite_master WHERE type='table';")
      stopifnot(nrow(tables) == 1)
      table_name <<- tables$name[1]
      print(paste("Tables in the database:", table_name))
      
      for (table in tables$name) {
        command <- paste("PRAGMA table_info(", table, ")", sep = "")
        attributes <- dbGetQuery(conn, command)
        print(paste("Table:", table))
        for (i in 1:nrow(attributes)) {
          attr <- attributes[i, ]
          print(paste("Attribute:", attr$name, ", Type:", attr$type, ", Default Val:", attr$dflt_value, ", Primary Key:", attr$pk))
        }
      }
    },
    close_connection = function() {
      if (!is.null(conn)) {
        dbDisconnect(conn)
      }
    },
    get_image = function(plate, well, channel) {
      if (is.numeric(well)) {
        query <- sprintf("SELECT path_unmixed FROM images WHERE plate = %d AND site = %d AND channel_num = %d", plate, well, channel)
      } else if (is.list(well)) {
        query <- sprintf("SELECT path_unmixed FROM images WHERE plate = %d AND row = '%s' AND column = %d AND field = %d AND channel_num = %d", 
                         plate, well[[1]], well[[2]], well[[3]], channel)
      }
      rows <- dbGetQuery(conn, query)
      stopifnot(nrow(rows) == 1)
      path <- file.path(data_root, rows$path_unmixed)
      
      if (grepl("\\.jpg$", path, ignore.case = TRUE)) {
        return(readJPEG(path))
      } else if (grepl("\\.png$", path, ignore.case = TRUE)) {
        return(readPNG(path))
      } else if (grepl("\\.tif$|\\.tiff$", path, ignore.case = TRUE)) {
        return(readTIFF(path))
      } else {
        stop("Unsupported image format")
      }
    },
    get_treatment = function(plate, well, channel) {
      if (is.numeric(well)) {
        query <- sprintf("SELECT treatment FROM images WHERE plate = %d AND site = %d AND channel_num = %d", plate, well, channel)
      } else if (is.list(well)) {
        query <- sprintf("SELECT treatment FROM images WHERE plate = %d AND row = '%s' AND column = %d AND field = %d AND channel_num = %d", 
                         plate, well[[1]], well[[2]], well[[3]], channel)
      }
      rows <- dbGetQuery(conn, query)
      stopifnot(nrow(rows) == 1)
      return(rows$treatment)
    },
    get_line = function(plate, well, channel) {
      if (is.numeric(well)) {
        query <- sprintf("SELECT line FROM images WHERE plate = %d AND site = %d AND channel_num = %d", plate, well, channel)
      } else if (is.list(well)) {
        query <- sprintf("SELECT line FROM images WHERE plate = %d AND row = '%s' AND column = %d AND field = %d AND channel_num = %d", 
                         plate, well[[1]], well[[2]], well[[3]], channel)
      }
      rows <- dbGetQuery(conn, query)
      stopifnot(nrow(rows) == 1)
      return(rows$line)
    }
  )
)

File: cellpaint_database/database_loader.py
import sqlite3
from sqlite3 import Error
from pathlib import Path
import numpy as np
from skimage import io


class DatabaseLoader:
    def __init__(self, db_file, data_root):
        self.db_file = db_file
        self.data_root = Path(data_root)
        self.conn = self.create_connection()

        cursor = self.conn.cursor()
        cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
        tables = cursor.fetchall()
        assert len(tables) == 1, "Database should only have one table"
        self.table_name = tables[0][0]
        print("Tables in the database:", tables)

        for table in tables:
            command = f"PRAGMA table_info({table[0]})"
            print(f"Table: {table}")
            cursor.execute(command)
            attributes = cursor.fetchall()
            for attr in attributes:
                print(
                    f"Attribute:{attr[1]}, Type:{attr[2]}, Default Val: {attr[4]}, Primary Key:{attr[5]}"
                )

    def create_connection(self):
        try:
            self.conn = sqlite3.connect(self.db_file)
            return self.conn
        except Error as e:
            print(e)

    def close_connection(self):
        if self.conn:
            self.conn.close()

    def get_image(self, plate, well, channel):
        cursor = self.conn.cursor()
        condition = self._get_where_from_well(plate, well, channel)

        command = f"SELECT path_unmixed FROM images WHERE {condition}"
        cursor.execute(command)
        rows = cursor.fetchall()
        assert len(rows) == 1, f"Expected 1 row, got {len(rows)}"
        path = rows[0][0]

        full_path = self.data_root / path

        return io.imread(full_path)

    def get_treatment(self, plate, well, channel):
        cursor = self.conn.cursor()
        condition = self._get_where_from_well(plate, well, channel)

        command = f"SELECT treatment FROM images WHERE {condition}"
        cursor.execute(command)
        rows = cursor.fetchall()
        assert len(rows) == 1, f"Expected 1 row, got {len(rows)}"
        treatment = rows[0][0]

        return treatment

    def get_line(self, plate, well, channel):
        cursor = self.conn.cursor()
        condition = self._get_where_from_well(plate, well, channel)

        command = f"SELECT line FROM images WHERE {condition}"
        cursor.execute(command)
        rows = cursor.fetchall()
        assert len(rows) == 1, f"Expected 1 row, got {len(rows)}"
        line = rows[0][0]

        return line
    
    def get_disease_type(self, plate, well, channel):
        cursor = self.conn.cursor()
        condition = self._get_where_from_well(plate, well, channel)

        command = f"SELECT disease_type FROM images WHERE {condition}"
        cursor.execute(command)
        rows = cursor.fetchall()
        assert len(rows) == 1, f"Expected 1 row, got {len(rows)}"
        dtype = rows[0][0]

        return dtype
    
    def _get_where_from_well(self, plate, well, channel):

        if type(well) == int:  # searching by site
            condition = f"plate = {plate} AND site = {well} AND channel_num = {channel}"
            
        elif type(well) == tuple:
            row, column, field = well
            condition = f"plate = {plate} AND row = '{row}' AND column = {column} AND field = {field} AND channel_num = {channel}"
            
        return condition

    
    def get_well_from_site(self, plate, site, image_session=None):
        if image_session is None:
            print("WARNING: assuming image_session = 1")
            image_session = 1

        cursor = self.conn.cursor()
        command = f"SELECT row, column, field FROM images WHERE plate = {plate} AND site = {site} AND image_session = {image_session}"
        cursor.execute(command)
        rows = cursor.fetchall()

        return rows[0]
